libcst/matchers/tests/test_extract.py

```diff
@@ -0,0 +1,411 @@
+            )
+            .args[1]
+            .value
+        )
+        self.assertEqual(nodes, {"arg": extracted_node})
+
+    def test_extract_precedence_sequence_wildcard(self) -> None:
+        expression = cst.parse_expression("a + b[c], d(e, f * g)")
+        nodes = m.extract(
+            expression,
+            m.Tuple(
+                elements=[
+                    m.DoNotCare(),
+                    m.Element(
+                        m.Call(
+                            args=[
+                                m.ZeroOrMore(
+                                    m.Arg(m.SaveMatchedNode(m.DoNotCare(), "arg"))
+                                ),
+                            ],
+                        )
+                    ),
+                ]
+            ),
+        )
+        extracted_node = (
+            cst.ensure_type(
+                cst.ensure_type(expression, cst.Tuple).elements[1].value, cst.Call,
+            )
+            .args[1]
+            .value
+        )
+        self.assertEqual(nodes, {"arg": extracted_node})
+
+    def test_extract_optional_wildcard(self) -> None:
+        expression = cst.parse_expression("a + b[c], d(e, f * g)")
+        nodes = m.extract(
+            expression,
+            m.Tuple(
+                elements=[
+                    m.DoNotCare(),
+                    m.Element(
+                        m.Call(
+                            args=[
+                                m.ZeroOrMore(),
+                                m.ZeroOrOne(
+                                    m.Arg(m.SaveMatchedNode(m.Attribute(), "arg")),
+                                ),
+                            ],
+                        )
+                    ),
+                ]
+            ),
+        )
+        self.assertEqual(nodes, {})
+
+    def test_extract_optional_wildcard_present(self) -> None:
+        expression = cst.parse_expression("a + b[c], d(e, f * g, h.i.j)")
+        nodes = m.extract(
+            expression,
+            m.Tuple(
+                elements=[
+                    m.DoNotCare(),
+                    m.Element(
+                        m.Call(
+                            args=[
+                                m.DoNotCare(),
+                                m.DoNotCare(),
+                                m.ZeroOrOne(
+                                    m.Arg(m.SaveMatchedNode(m.Attribute(), "arg")),
+                                ),
+                            ],
+                        )
+                    ),
+                ]
+            ),
+        )
+        extracted_node = (
+            cst.ensure_type(
+                cst.ensure_type(expression, cst.Tuple).elements[1].value, cst.Call,
+            )
+            .args[2]
+            .value
+        )
+        self.assertEqual(nodes, {"arg": extracted_node})
+
+    def test_extract_sequence(self) -> None:
+        expression = cst.parse_expression("a + b[c], d(e, f * g, h.i.j)")
```

Comment:
Probably my newness makes this hard to read, but if I'm understanding correctly the input/match/assert here is the same thing? Would it be reasonable to wrap a helper that allows specifying the input and output as an expression for a particular matcher? I suppose variable names would get in the way.

Reply:
Its pretty samey and I'll admit to copy-pastaing the tests a bit. I'm on the fence about improving their quality past the point of validating the featureset though.